### PR TITLE
Data virtualization

### LINKED ISSUE #29: Data virtualization
Don't read all flies on scene loading. Create visualization frames objects asynchronously on scrolling and indicate loading.

### PR DESCRIPTION
**Pull request by Alexander-Ploskin**

### CODE REVIEW

File: src/main/kotlin/sliv/tool/scene/SceneFacade.kt [hunk 1]
```diff
@@ -31,49 +36,34 @@ class SceneFacade(private val controller: SceneController) {
...
-                ?: throw IllegalStateException("No visualization layer is created for ${file.projectLayer.name}")
-            landmarks[visualizationLayer] = createLandmarks(file).toList()
+        val getImage = { Image(FileInputStream(imagePath.toFile())) }
+
+        val getLandmarks = {
+            val landmarks = HashMap<Layer, List<Landmark>>()
+            landmarkFiles.forEach { file ->
+                val visualizationLayer = visualizationLayers[file.projectLayer.name]
+                    ?: throw IllegalStateException("No visualization layer is created for ${file.projectLayer.name}")
+                landmarks[visualizationLayer] = createLandmarks(file).toList()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/SceneFacade.kt]: why extra `toList `is required?

File: src/main/kotlin/sliv/tool/scene/SceneFacade.kt [hunk 2]
```diff
@@ -31,49 +36,34 @@ class SceneFacade(private val controller: SceneController) {
...
         }
-        return VisualizationFrame(null, this.timestamp, landmarks.toMap())
+
+        return VisualizationFrame(this.timestamp, getImage, getLandmarks)
     }
 
-    //TODO: use real parser instead of test data
-    private fun createLandmarks(file: LandmarkFile): Sequence<Landmark> {
+    private fun createLandmarks(file: LandmarkFile): List<Landmark> {
         val layer = visualizationLayers[file.projectLayer.name]!!
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/SceneFacade.kt]: provide this to func, you have checked it before

File: src/main/kotlin/sliv/tool/scene/model/VisualizationFrame.kt [hunk 3]
```diff
@@ -2,5 +2,16 @@ package sliv.tool.scene.model
...
-//TODO: null Image have to be removed after data virtualization will be done
-class VisualizationFrame(val image: Image?, val timestamp: Long, val landmarks: Map<Layer, List<Landmark>>)
\ No newline at end of file
+class VisualizationFrame(
+    val timestamp: Long,
+    private val getImage: () -> Image,
+    private val getLandmarks: () -> Map<Layer, List<Landmark>>
+) {
+    val image: Image
+        get() {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/model/VisualizationFrame.kt]: `get() = getImage()`

File: src/main/kotlin/sliv/tool/scene/view/FrameView.kt [hunk 4]
```diff
@@ -31,25 +37,51 @@ class FrameView(
     }
 
     fun setFrame(frame: VisualizationFrame) {
-        //TODO: warn if image size doesn't equal to the frame size
+        currentJob?.cancel()
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/FrameView.kt:52]: will it be cancelled if reimport files to scene?

File: src/main/kotlin/sliv/tool/scene/view/FrameView.kt [hunk 5]
```diff
@@ -31,25 +37,51 @@ class FrameView(
...
+        draw()
+    }
+
+    private fun reloadData(frame: VisualizationFrame) {
         landmarksViews = frame.landmarks.mapValues {
             it.value.map { landmark -> LandmarkView.create(landmark) }
         }
-        draw()
+        image = frame.image
+        if(image!!.height != height || image!!.width != width) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/FrameView.kt]: remove `!!`

File: src/main/kotlin/sliv/tool/scene/view/FrameView.kt [hunk 6]
```diff
@@ -31,25 +37,51 @@ class FrameView(
...
+            return
+        }
+
+        gc.drawImage(image, 0.0, 0.0, canvas.width, canvas.height)
         doForAllEnabledLandmarks { view -> view.draw(gc, scale.value) }
     }
 
-    private fun drawFakeImage() {
+    private val isDataLoaded
+        get() = landmarksViews != null && image != null
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/FrameView.kt:90]: are these variables assigned in other thread? maybe volatile or thread switch assignment?

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt [hunk 7]
```diff
@@ -30,19 +30,18 @@ class SceneView : View() {
         root.children.clear()
 
         val scene = controller.scene.value
-        val frames = (0 until scene.framesCount).mapNotNull { i -> scene.getFrame.invoke(i) }
 
-        //TODO: get size from the first image after data virtualization will be done
-        val width = 500.0
-        val height = 500.0
+        val firstImage = scene.frames.first().image
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/SceneView.kt:45]: what if no images provided? `firstOrNull`?

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: src/main/kotlin/sliv/tool/scene/view/FrameView.kt (hunk 4 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/FrameView.kt:52]: I decided to create coroutine scope in the scene view and cancel it if data was reset

**Review by DmiitriyJarosh — APPROVED** [review]

**Pull request merged**